keps/sig-cloud-provider/20191004-out-of-tree-credential-providers.md

```diff
@@ -0,0 +1,207 @@
+
+In addition to supporting cloud provider migration this KEP supports allowing users to support multiple container registries across different cloud providers by introducing a pluggable interface to cloud specific credential providers such that a single user could run multiple credential providers for different clouds at the same time. Currently this functionality is gated by having only a single, active cloud provider within the API server and kubelet.
+
+### Goals
+
+* Develop/test/release an interface for kubelet to obtain registry credentials from a cloud provider specific binary
+* Update/test/release the credential acquisition logic within kubelet
+* Build user documentation for out-of-tree credential providers
+* Migrate existing in-tree credential providers to new credential provider interface
+* Remove in-tree credential provider code from kuberentes core
+
+### Non-Goals
+
+* Broad removal of cloud SDK usage falls under the [KEP for removing in-tree providers](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cloud-provider/2019-01-25-removing-in-tree-providers.md).
+* Continuing to support projects that import the credential provider package.
+
+## Proposal
+
+### External Credential Provider
+
+An executable capable of providing container registry credentials will be installed on each node. This binary will be executed by the kubelet to obtain container registry credentials in a format compatible with container runtimes. Credential responses may be cached within the kubelet.
+
+This architecture is similar to the approach taken by the exec based credential plugin architecture already present in client-go, and CNI and is a well understood pattern.  The API types are modeled after the ExecConfig and ExecCredential in client-go which define exec based credential retrieval for similar use cases.
+
+A RegistryCredentialProvider configuration API type (similar to [clientauthentication](https://github.com/kubernetes/kubernetes/tree/0273d43ae9486e9d0be292c01de2dd4143522b86/staging/src/k8s.io/client-go/pkg/apis/clientauthentication/v1beta1))will be added to Kubernetes:
+
+```go
+// RegistryCredentialProvider is used by the kubelet container runtime to match image URIs with
+// exec-based credential provider plugins that provide container registry credentials.
+type RegistryCredentialProvider struct {
+    metav1.TypeMeta `json:",inline"`
+
+    // ImageURLMatch is used to match against the Image URL to determine which images to
+    // provide credentials for.
+    //
+    // The ImageURLMatch string is a URL pattern which can optionally contain a port and a path.  Globs can be used in the hostname (but not the port or the path).
```

Comment:
You keep referring to string but ImageURLMatch appears to be an array of strings. If it meant to be an array it would be nice to talk about how the set behaves. Eg. Does only one URL need to match? Also for arrays it would be nice to have the name be plural. ImageURLMatchers?

Reply:
Made clearer, yes it only needs one URL to match.